c/include/libsbp/sbp.h

```diff
@@ -23,39 +23,91 @@ extern "C" {
 /** Return value indicating a CRC error. */
-#define SBP_CRC_ERROR      -2
+#define SBP_CRC_ERROR            -2
 /** Return value indicating an error occured whilst sending an SBP message. */
-#define SBP_SEND_ERROR     -3
+#define SBP_SEND_ERROR           -3
 /** Return value indicating an error occured because an argument was NULL. */
-#define SBP_NULL_ERROR     -4
+#define SBP_NULL_ERROR           -4
 /** Return value indicating an error occured in the write() operation. */
-#define SBP_WRITE_ERROR     -5
+#define SBP_WRITE_ERROR          -5
 /** Return value indicating an error occured in the read() operation. */
-#define SBP_READ_ERROR     -6
+#define SBP_READ_ERROR           -6
+/** Return value indicating too many bytes were read in read() operation */
+#define SBP_READ_TOO_MANY_ERROR  -7
 
 /** Default sender ID. Intended for messages sent from the host to the device. */
 #define SBP_SENDER_ID 0x42
+/** Header length in bytes. */
+#define SBP_HEADER_LEN 6
+/** CRC length in bytes. */
+#define SBP_CRC_LEN 2
+/** Max payload length in bytes. */
+#define SBP_MAX_PAYLOAD_LEN (UINT8_MAX)
+/** Max frame length in bytes. */
+#define SBP_MAX_FRAME_LEN (SBP_HEADER_LEN + SBP_MAX_PAYLOAD_LEN + SBP_CRC_LEN)
 
-/** SBP callback function prototype definition. */
+/** Frame offset for the preamble. */
+#define SBP_FRAME_OFFSET_PREAMBLE (0u)
+/** Frame offset for the preamble. */
+#define SBP_FRAME_OFFSET_MSGTYPE (SBP_FRAME_OFFSET_PREAMBLE + sizeof(u8))
+/** Frame offset for the preamble. */
+#define SBP_FRAME_OFFSET_SENDERID (SBP_FRAME_OFFSET_MSGTYPE + sizeof(u16))
+/** Frame offset for the preamble. */
+#define SBP_FRAME_OFFSET_MSGLEN (SBP_FRAME_OFFSET_SENDERID + sizeof(u16))
+/** Frame offset for the preamble. */
+#define SBP_FRAME_OFFSET_MSG (SBP_FRAME_OFFSET_MSGLEN + sizeof(u8))
+/** Frame offset for the CRC is a function of msg length. */
+#define SBP_FRAME_OFFSET_CRC(msg_len) (SBP_FRAME_OFFSET_MSG + msg_len)
+#define SBP_FRAME_CALC_LEN(msg_len) (SBP_HEADER_LEN + msg_len + SBP_CRC_LEN)
+
+/** Get message payload pointer from frame */
+#define SBP_FRAME_MSG_PAYLOAD(frame_ptr) (&(frame_ptr[SBP_FRAME_OFFSET_MSG]))
+
+/** SBP_MSG_ID to use to register frame callback for ALL messages. */
+#define SBP_MSG_ALL 0
+
+/** SBP callback function prototype definitions. */
 typedef void (*sbp_msg_callback_t)(u16 sender_id, u8 len, u8 msg[], void *context);
+typedef void (*sbp_frame_callback_t)(u16 sender_id, u16 msg_type,
+                                     u8 payload_len, u8 payload[],
+                                     u16 frame_len, u8 frame[], void *context);
+
+/** SBP callback type enum:
+ * SBP_TYPE_CALLBACK are the original callbacks in libsbp without framing args
+ * SBP_TYPE_FRAME are raw frame callbacks that include framing data as args.
+ * This enum is stored on each sbp_msg_callback_node struct to identify how
+ * to cast the callback function pointer stored within.
+ */
+enum sbp_cb_type {
+  SBP_TYPE_CALLBACK = 0,
```

Comment:
I am starting to think this should be SBP_PAYLOAD_CALLBACK instead of SBP_TYPE_CALLBACK.  I had named it type callback when it was impossible to register a frame callback on a particular message type.